packages/eslint-plugin/docs/rules/no-duplicate-type-constituents.md

```diff
@@ -0,0 +1,64 @@
+---
+description: 'Disallow duplicate constituents of union or intersection types.'
+---
+
+> 🛑 This file is source code, not the primary documentation location! 🛑
+>
+> See **https://typescript-eslint.io/rules/no-duplicate-type-constituents** for documentation.
+
+Although TypeScript supports duplicate union and intersection constituents, people usually expect members to have unique values within the same intersection and union. Duplicate values make the code redundant and generally reduce readability.
+
+## Rule Details
+
+This rule disallows duplicate union or intersection constituents.
+
+It determines whether two types are equivalent in the following way.
+
+1. whether the syntax is exactly the same.
+2. whether TypeScript treats them as the same type.
+
+If either of the two conditions is satisfied, It treats the two types as duplicates.
```

Comment:
I don't think most users really care about how we determine the equivalency internally. The syntax being exactly the same is an internal optimization to avoid using the type checker if we don't have to.

```suggestion
We consider types to be duplicate if they evaluate to the same result in the type system.
For example, given `type A = string` and `type T = string | A`, this rule would flag `string | A`.
```

Reply:
Certainly, there was no need to even talk about internal optimization here...